File: SwrveSDK/SDK/Messaging/SwrveTrigger.m
#import "SwrveTrigger.h"
#import "SwrveTriggerCondition.h"

@implementation SwrveTrigger

@synthesize eventName = _eventName;
@synthesize conditions = _conditions;
@synthesize isValidTrigger = _isValidTrigger;

#pragma mark - public function

+ (NSArray *) initTriggersFromDictionary:(NSDictionary *)dictionary {
    
    NSArray* jsonTriggers = [dictionary objectForKey:kTriggerEventListKey];
    if(!jsonTriggers){
        return nil;
    }

    NSMutableArray *resultantTriggers = [NSMutableArray new];
    for (NSDictionary* trigger in jsonTriggers){
        if (trigger) {
            SwrveTrigger *swrveTrigger = [[SwrveTrigger alloc] initWithDictionary:trigger];
            if(swrveTrigger) {
                [resultantTriggers addObject:swrveTrigger];
            }
        }
    }
    return resultantTriggers;
}

#pragma mark - private functions

- (id) initWithDictionary:(NSDictionary *)dictionary {
    self = [super init];
    if(self) {
        _isValidTrigger = YES;
        _eventName = [[dictionary objectForKey:kTriggerEventNameKey] lowercaseString];
        _conditions = [self produceConditionsFromDictionary: [dictionary objectForKey:kTriggerEventConditionsKey]];
        if(_isValidTrigger){
            return self;
        }
    }
    return nil;
}

- (NSArray *) produceConditionsFromDictionary:(NSDictionary *) dictionary {
    
    NSMutableArray *resultantConditions = [NSMutableArray new];
    NSString *triggerOperator = [dictionary objectForKey:@"op"];
    
    if(![dictionary count]){
        return nil;
    }
    
    if([triggerOperator isEqualToString:@"eq"] || [triggerOperator isEqualToString:@"contains"]) {

        SwrveTriggerCondition *condition = [[SwrveTriggerCondition alloc] initWithDictionary:dictionary andOperator:nil];
        if(condition) {
            [resultantConditions addObject:condition];
        } else {
            _isValidTrigger = NO;
            return nil;
        }
    } else if([triggerOperator isEqualToString:@"and"]) {

        NSDictionary *arguments = [dictionary objectForKey:@"args"];
        if(!arguments){
            _isValidTrigger = NO;
            return nil;
        }
        
        for(NSDictionary *triggerCondition in arguments) {
            
            SwrveTriggerCondition *condition = [[SwrveTriggerCondition alloc] initWithDictionary:triggerCondition andOperator:triggerOperator];
            if (condition) {
                [resultantConditions addObject:condition];
            }
        }
    } else if([triggerOperator isEqualToString:@"or"]) {
        
        NSDictionary *arguments = [dictionary objectForKey:@"args"];
        if(!arguments){
            _isValidTrigger = NO;
            return nil;
        }
        
        for(NSDictionary *triggerCondition in arguments) {
            
            SwrveTriggerCondition *condition = [[SwrveTriggerCondition alloc] initWithDictionary:triggerCondition andOperator:triggerOperator];
            if (condition) {
                [resultantConditions addObject:condition];
            }
        }
    } else {
        _isValidTrigger = NO;
        return nil;
    }
    
    for (SwrveTriggerCondition *condition in resultantConditions) {
        
        switch (condition.triggerOperator) {
            case SwrveTriggerOperatorAND:
                if([resultantConditions count] <= 1){
                    _isValidTrigger = NO;
                }
                break;
            case SwrveTriggerOperatorOR:
                if([resultantConditions count] < 1){
                    _isValidTrigger = NO;
                }
                break;
            case SwrveTriggerOperatorOTHER:
                if([resultantConditions count] > 1){
                    _isValidTrigger = NO;
                }
                break;
            default:
                break;
        }
        
        switch (condition.conditionOperator) {
            case SwrveTriggerOperatorEQUALS:
                break;
            case SwrveTriggerOperatorCONTAINS:
                break;
            case SwrveTriggerOperatorOTHER:
                _isValidTrigger = NO;
                break;
            default:
                _isValidTrigger = NO;
                break;
        }
    }
    
    return resultantConditions;
}

- (BOOL) canTriggerWithPayload:(NSDictionary *)payload {
    
    BOOL canTrigger = YES;
    
    if([_conditions count] > 0) {
        
        for (SwrveTriggerCondition *condition in _conditions) {
            
            canTrigger = [condition hasFulfilledCondition:payload];
            
            if (condition.triggerOperator == SwrveTriggerOperatorOR) {
                if (canTrigger) {
                    return true;
                }
            } else if (condition.triggerOperator == SwrveTriggerOperatorAND)  {
                if(!canTrigger) {
                    return NO;
                }
            } else {
                return canTrigger;
            }
        }
    }
    
    return canTrigger;
}

- (BOOL) isValidTrigger {
    return _isValidTrigger;
}

@end

File: SwrveSDKSwift/SwrveTrigger.swift
import Foundation

@objc public class SwrveTrigger: NSObject {

    static let triggerEventListKey = "triggers"
    let triggerEventNameKey = "event_name"
    let triggerEventConditionsKey = "conditions"

    /// Event name for the trigger
    @objc public private(set) var eventName: String

    /// Conditions for the trigger
    @objc public var conditions: [SwrveTriggerCondition]

    /// Indicates if this is a valid trigger
    @objc public var isValidTrigger: Bool = true

    /// Initializes triggers from a dictionary
    /// - Parameter dictionary: The dictionary containing trigger information
    class func initTriggers(from dictionary: [String: Any]) -> [SwrveTrigger]? {
        guard let jsonTriggers = dictionary[triggerEventListKey] as? [[String: Any]] else {
            return nil
        }
        var resultantTriggers = [SwrveTrigger]()
        for triggerData in jsonTriggers {
            if let swrveTrigger = SwrveTrigger(dictionary: triggerData) {
                resultantTriggers.append(swrveTrigger)
            }
        }
        return resultantTriggers
    }

    /// Initializes a trigger from a dictionary
    /// - Parameter dictionary: The dictionary containing trigger information
    @objc public init?(dictionary: [String: Any]) {
        guard let eventName = dictionary[triggerEventNameKey] as? String else {
            return nil
        }

        self.eventName = eventName.lowercased()
        self.conditions = []
        super.init()

        if let conditionData = dictionary[triggerEventConditionsKey] as? [String: Any] {
            self.conditions = produceConditions(from: conditionData) ?? []
        }

        if !self.isValidTrigger {
            return nil
        }
    }

    // MARK: - Private Methods

    /// Produces conditions from a dictionary
    /// - Parameter dictionary: The dictionary containing condition information
    private func produceConditions(from dictionary: [String: Any]) -> [SwrveTriggerCondition]? {
        var resultantConditions = [SwrveTriggerCondition]()
        guard !dictionary.isEmpty else {
            return nil
        }
        let triggerOperator: String = dictionary["op"] as? String ?? ""

        if ["eq", "contains", "number_gt", "number_eq", "number_lt", "number_between", "number_not_between"].contains(triggerOperator) {
            if let condition = SwrveTriggerCondition(dictionary: dictionary, operatorKey: nil) {
                resultantConditions.append(condition)
            } else {
                isValidTrigger = false
                return nil
            }

        } else if triggerOperator == "and" || triggerOperator == "or" {
            guard let arguments = dictionary["args"] as? [[String: Any]] else {
                isValidTrigger = false
                return nil
            }
            for triggerConditionData in arguments {
                if let condition = SwrveTriggerCondition(dictionary: triggerConditionData, operatorKey: triggerOperator) {
                    resultantConditions.append(condition)
                }
            }
        } else {
            isValidTrigger = false
            return nil
        }

        for condition in resultantConditions {
            switch condition.triggerOperator {
            case .and:
                if resultantConditions.count <= 1 {
                    isValidTrigger = false
                }
            case .or:
                if resultantConditions.isEmpty {
                    isValidTrigger = false
                }
            case .other:
                if resultantConditions.count > 1 {
                    isValidTrigger = false
                }

            default:
                break
            }

            switch condition.conditionOperator {
            case .other:
                isValidTrigger = false
            default:
                break
            }
        }

        return resultantConditions
    }

    /// Checks if the trigger can fire based on a payload
    /// - Parameter payload: The payload containing event information
    /// - Returns: Boolean indicating if the trigger can fire
    @objc public func canTrigger(withPayload payload: [AnyHashable: Any]) -> Bool {
        var canTrigger = true

        if !conditions.isEmpty {
            for condition in conditions {
                canTrigger = condition.hasFulfilledCondition(with: payload)

                if condition.triggerOperator == .or {
                    if canTrigger { return true }
                } else if condition.triggerOperator == .and {
                    if !canTrigger { return false }
                } else {
                    return canTrigger
                }
            }
        }

        return canTrigger
    }

    /// Checks if this is a valid trigger
    /// - Returns: Boolean indicating if the trigger is valid
    func isValid() -> Bool {
        isValidTrigger
    }
}
